### DISCUSSION TITLE
Performance impact of large number of collections

### DISCUSSION THREAD

**Discussion by jubingc** [milvus-io/milvus#37594] [Q&A and General discussion]
According to the [documentation](https://milvus.io/docs/limit_collection_counts.md):

> A Milvus instance allows up to 65,536 collections. However, too many collections may result in performance issues.

I would like to better understand how the number of collections and partitions affects performance. Could you clarify what constitutes "too many" collections in this context?

As an example, the calculation below multiplies the number of collections, shards, and partitions. However, [shards are primarily for data writing, while partitions and segments are used for data reading](https://zilliz.com/blog/sharding-partitioning-segments-get-most-from-your-database). Why are these elements multiplied together?

> 60 (collections) x 2 (shards) x 4 (partitions) + 40 (collections) x 1 (shard) x 12 (partitions) = 960

Additionally, per the [documentation](https://milvus.io/docs/limitations.md), the maximum number of partitions in a collection is 4,096 (with a default of 1,024, controlled by [rootCoord.maxPartitionNum](https://milvus.io/docs/configure_rootcoord.md#rootCoordmaxPartitionNum)). Given a shared [rootCoord.maxGeneralCapacity](https://milvus.io/docs/configure_rootcoord.md#rootCoordmaxGeneralCapacity), which of the following configurations would likely yield better performance?

1. 1,024 collections, 2 shards per collection, 16 partitions per collection = 32,768 general capacity
2. 16 collections, 2 shards per collection, 1,024 partitions per collection = 32,768 general capacity

Beyond performance, I’d also appreciate insights into the pros and cons of each setup. Some drawbacks of the second setup I’m aware of include: 
a. The recommended size for a partition is up to 1 billion items ([reference](https://milvus.io/docs/manage-partitions.md)). 
b. There is currently no way to filter data within a partition quickly.

Are there additional pros or cons to consider for each of these configurations?

**Comment by yhmo** [comment]
Take a look at this chart so understand how milvus manages the data in shards/partitions/collections/segments:
![channels](https://github.com/user-attachments/assets/1ae5827a-f1fd-4085-b67c-938b1fb25f0b)

**Reply by yhmo** [reply]
Assume a collection has 2 shards, 2 partitions. To maintain the data in this collection, it maintains 2 virtual-channels for each partition. So, there will be 2 * 2 virtual channels to be managed. If there are 60 collections, and each collection has 2 shards, 4 partitions, there will be 60 * 2 * 4 virtual channel objects to be managed. Each single row of data passed via a virtual channel needs to be carefully recorded.

**Reply by yhmo** [reply]
1,024 collections, 2 shards per collection, 16 partitions
16 collections, 2 shards per collection, 1,024 partitions

No much difference between the two cases, since there will be 1024 * 2 * 16 virtual channels need to be maintained.

Typically, it is recommended that the total number of virtual channels be controlled under 1000. In the v2.4.x, maybe v-channels number 5000 ~ 10000 also can work. Anyway, we don't recommend high-number v-channels.

**Reply by jubingc** [reply]
@yhmo thanks. That diagram is helpful. Is it available somewhere in the doc? Suppose we have more than 10K v-channel, which component would become the bottleneck first?

**Reply by jubingc** [reply]
@yhmo v-channel is for write performance, right? How about the read performance, do the following two options have the same read performance?

1. 1,024 collections, 2 shards per collection, 16 partitions
2. 16 collections, 2 shards per collection, 1,024 partitions


I have this confusion because of this sentence from the [doc](https://milvus.io/docs/multi_tenancy.md)

> The search performance of partition-oriented multi-tenancy is much better than collection-oriented multi-tenancy.

Why does partition-oriented multi-tenancy have better search performance than collection-oriented multi-tenancy? @xiaofan-luan

**Reply by yhmo** [reply]
As @xiaofan-luan mentioned "partition is considered to be more light weight than collection". So, partition-oriented is better.

**Comment by xiaofan-luan** [comment]
@yanliang567 is working on the effect of large number of the collections/partitions

The goal here is to support:
10000 collections with 4096 partitions 
in oue cluster.

This could be part of milvus 2.5.X

**Reply by jubingc** [reply]
@xiaofan-luan, regarding your mention that Milvus 2.5.X will support:

> 10,000 collections with 4,096 partitions in one milvus cluster

Does this imply that the total number of v-channels will be 10,000 * 4,096? From @yhmo's response above, it seems that the number of v-channels significantly impacts performance. If so, the performance calculation would depend primarily on the number of v-channels, determined by multiplying the number of partitions by the number of shards. Collections, by contrast, don’t appear to directly impact performance, except that partitions and shards are contained within individual collections and cannot cross collection boundaries.

Also, is there an estimated release timeline for Milvus 2.5.X?

**Reply by xiaofan-luan** [reply]
collectition * shard * partition is the least segment size you get for a cluster.
For 1000 collection * 2 shard * 1000 partition, you get 2M segments (which could be huge and brings a lot of head on maintain it)
here is some trick:
1. for small collection(< 100m), only one shard is required
2.partition is considered to be more light weight than collection
3. But the partition number per collection is limited, we are gonna to test it, but right now our recommendation is less than 10K

So something we can try is to start from 16 collection * 1 shard * 1000 partition. This should work perfectly.

**Reply by xiaofan-luan** [reply]
2.5 is gonna to released soon this week or next. We don't have those improvements yet. this is currently under evaluation and need to be improved

**Reply by jubingc** [reply]
@xiaofan-luan Thank you for the suggestion.

Beyond performance, we're also evaluating strategies for multi-tenancy. Our use case involves managing Milvus as a service for internal teams. Currently, we use collections for multi-tenancy, assigning each user to their own collection since usage patterns vary, and we can’t predict the collection sizes—users may ingest data unpredictably. Currently, we have around 150 collections, each with only one partition. All collections do not share the same schema, so partition-based multi-tenancy might not be an option.

If we were to use partitions for multi-tenancy, though, we’d likely encounter some limitations sooner, including:

1. All partitions in the same collection must share the same schema.
2. The maximum partition limit per collection (4,096) is significantly lower than the maximum number of collections (65,536).
3. Data filtering by partition would no longer be an option (a tenant could have multiple partitions, which adds management overhead).
4. Partition size limitations (1B records) could cause users to reach size limits earlier than if they used collections.

Any suggestions would be helpful. Thanks!

**Reply by xiaofan-luan** [reply]
> @xiaofan-luan Thank you for the suggestion.
> 
> Beyond performance, we're also evaluating strategies for multi-tenancy. Our use case involves managing Milvus as a service for internal teams. Currently, we use collections for multi-tenancy, assigning each user to their own collection since usage patterns vary, and we can’t predict the collection sizes—users may ingest data unpredictably. Currently, we have around 150 collections, each with only one partition. All collections do not share the same schema, so partition-based multi-tenancy might not be an option.
> 
> If we were to use partitions for multi-tenancy, though, we’d likely encounter some limitations sooner, including:
> 
> 1. All partitions in the same collection must share the same schema.
> 2. The maximum partition limit per collection (4,096) is significantly lower than the maximum number of collections (65,536).
> 3. Data filtering by partition would no longer be an option (a tenant could have multiple partitions, which adds management overhead).
> 4. Partition size limitations (1B records) could cause users to reach size limits earlier than if they used collections.
> 
> Any suggestions would be helpful. Thanks!

How many tenants are you trying to work with?
If you tenant number is definitely less than 10000 then I would prefer to create multiple collections .
Right now we believe create 10000 colleection is acually doable at this moment, We do have a campaign to improve the collection number support.

**Reply by xiaofan-luan** [reply]
Would like to get more info about you are building and how we can help. Ping me at james.luan@zilliz.com.

Also want to share about some of the experiment we did during the test

**Reply by jubingc** [reply]
@xiaofan-luan  Thank you. I will be contacting you via email for further discussion.

We don't expect to have more than 4,000 tenants/collections in a cluster. In this case, we can adjust the number of partitions for specific tenants who require higher throughput.

**Reply by mihailyanchev** [reply]
> @yanliang567 is working on the effect of large number of the collections/partitions
> 
> The goal here is to support: 10000 collections with 4096 partitions in oue cluster.
> 
> This could be part of milvus 2.5.X

Quick question: When you refer to the total number of collections/partitions limits, do you mean the ones loaded in memory at a given time or in total - loaded and unloaded? Also I assume you mean to say this about cluster mode Milvus?

I'm asking because I implemented a sharding strategy where I keep the number of collections fixed, but allocate data separated by business logic in dedicated partitions. I load partitions in memory only on demand and unload all data from memory in a nightly pipeline job to reduce dramatically memory utilisation. So now I am wondering if my whole premise was wrong. As far as I understand, if I keep the total number of partitions in memory at a given time around some reasonable number (~5000) it is ok. Am I correct?

**Reply by yhmo** [reply]
> > @yanliang567 is working on the effect of large number of the collections/partitions
> > The goal here is to support: 10000 collections with 4096 partitions in oue cluster.
> > This could be part of milvus 2.5.X
> 
> Quick question: When you refer to the total number of collections/partitions limits, do you mean the ones loaded in memory at a given time or in total - loaded and unloaded? Also I assume you mean to say this about cluster mode Milvus?
> 
> I'm asking because I implemented a sharding strategy where I keep the number of collections fixed, but allocate data separated by business logic in dedicated partitions. I load partitions in memory only on demand and unload all data from memory in a nightly pipeline job to reduce dramatically memory utilisation. So now I am wondering if my whole premise was wrong. As far as I understand, if I keep the total number of partitions in memory at a given time around some reasonable number (~5000) it is ok. Am I correct?

I suppose the latest v2.5.4 can handle more collections/partitions than old versions. 5000 partitions should be ok.

" load partitions in memory only on demand and unload all data from memory in a nightly pipeline job" ------------------ It seems your business app search/query a part of collections/partitions every day? I think this solution is ok.

Related discussions:
https://github.com/milvus-io/milvus/discussions/33661
https://github.com/milvus-io/milvus/discussions/39600

**Reply by mihailyanchev** [reply]
Thanks for the info and resources!

Yes, the app allows search/query throughout the day and I implemented loading partitions upon demand.

**Reply by mihailyanchev** [reply]
@yhmo @xiaofan-luan Is there anything new here in terms of limits of collections x partitions in Milvus 2.6?

Also, is there a place in the docs where I can check how this topic evolves.

**Reply by yhmo** [reply]
> @yhmo @xiaofan-luan Is there anything new here in terms of limits of collections x partitions in Milvus 2.6?
> 
> Also, is there a place in the docs where I can check how this topic evolves.

Well, it is not like a "goal", it is something like "continuous improvement". Hopefully, it can handle 10k collections in the latest version of v2.5(v2.6 is similar). In each minor version, the internal machinery and memory/cpu usage are optimized so that it can hold more and more collections/partitions. However, the upper limit may be different in different scenarios. For example, 10k collections with static data(5000 entities) per collection, it work fine. But if you have 10k collections 1M entities per collection, and continuously insert/delete, it could run into trouble.
